contracts/RMRK/extension/emotable/RMRKEmotable.sol

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.16;
+
+import "./IRMRKEmotable.sol";
+
+/**
+ * @title RMRKEmotable
+ * @author RMRK team
+ * @notice Smart contract of the RMRK Emotable module.
+ */
+abstract contract RMRKEmotable is IRMRKEmotable {
+    // Used to avoid double emoting and control undoing
+    mapping(address => mapping(uint256 => mapping(bytes4 => uint256)))
+        private _emotesPerAddress;  // Cheaper than using a bool
+    mapping(uint256 => mapping(bytes4 => uint256)) private _emotesPerToken;
+
+    event Emoted(
+        address indexed emoter,
+        uint256 indexed tokenId,
+        bytes4 emoji,
+        bool on
+    );
+
+    /**
+     * @inheritdoc IRMRKEmotable
+     */
+    function getEmoteCount(
+        uint256 tokenId,
+        bytes4 emoji
+    ) public view returns (uint256) {
+        return _emotesPerToken[tokenId][emoji];
+    }
+
+    /**
+     * @notice Used to emote or undo an emote on a token.
+     * @param tokenId ID of the token being emoted
+     * @param emoji Unicode identifier of the emoji
+     * @param state whether to turn emote or undo. True for emote, false for undo
+     */
+    function _emote(
+        uint256 tokenId,
+        bytes4 emoji,
+        bool state
+    ) internal virtual {
+        bool currentVal = _emotesPerAddress[msg.sender][tokenId][emoji] == 1;
+        if (currentVal != state) {
+            _beforeEmote(tokenId, emoji, state);
+            if (state) {
+                _emotesPerToken[tokenId][emoji] += 1;
+            } else {
+                _emotesPerToken[tokenId][emoji] -= 1;
+            }
+            _emotesPerAddress[msg.sender][tokenId][emoji] = state ? 1 : 0;
+            emit Emoted(msg.sender, tokenId, emoji, state);
+        }
```

Comment:
I think this one could use an `else` clause:

```suggestion
        if (currentVal != state) {
            _beforeEmote(tokenId, emoji, state);
            if (state) {
                _emotesPerToken[tokenId][emoji] += 1;
            } else {
                _emotesPerToken[tokenId][emoji] -= 1;
            }
            _emotesPerAddress[msg.sender][tokenId][emoji] = state ? 1 : 0;
            emit Emoted(msg.sender, tokenId, emoji, state);
        } else {
            revert RMRKAlreadyEmoted();
        }
```

Reply:
It was an option, but I don't see the harm on trying to reapply a state so I went for no revert.
Additionally, in this proposal you would raise `RMRKAlreadyEmoted` when state is off and you're setting off again. So the error doesn't make sense, we'd need to errors (more code and contract size) or a more generic. In any case I don't see a value against just ignoring.

Reply:
I think no revert path is good if we are certain that the function will be called directly and that no other actions will be taken afterwards. If there is a relatively high possibility of other on chain actions following this one in the same transaction, I think it would be useful to revert in order to use as little gas as possible. I agree about the error having to be more generic.

That said, I'm ok with the current implementation if you both agree with it 🦾

Reply:
If it's going to be called before something else (totally possible) the caller shouldn't assume such behavior. What if we just add to the docstring that will do nothing if state is already the same?  
@stevyhacker  what are your thought's on this?

Reply:
I am okay with the current implementation, emote function works like a toggle button in this case, no real benefit for raising an error/reverting here as this is a very simple operation.

Reply:
Let's add `@dev` tag noting this down like you suggested @steven2308 and finalize this discussion 👌